### ISSUE TITLE
[BUG] Avoid building all the providers to lessen CI load

### ISSUE THREAD

**Issue by vados-cosmonic** [wasmCloud/wasmCloud#3481]
### Affected project(s)

- [ ] documentation
- [ ] examples
- [ ] wasmCloud host
- [X] wasmCloud CLI (wash)
- [ ] wasmCloud dashboard UI (washboard)
- [ ] capability providers
- [ ] provider bindgen
- [ ] control interface client
- [X] other / not sure

### Describe the bug

At present, the CI run for jobs like `oci` require a large amount of memory to run. This has a few causes, but it seems that chief of them is that we build the providers (and every package) at the same time when we do not need to (for example to build an OCI image that only has a subset of packages inside)

It'd be ideal if we could refactor either our regular build or CI to build *only* the required packages (i.e. `wash`, not `provider-http-server`).

There are a bunch of ways to solve this:

- Remove the provider build setup in `Cargo.toml`, and refactor the codebase some other way to achieve the same goals (ensuring providers are built/easily available to tests/etc)
- Modify the existing workflow to ensure that other packages are not built when `oci` runs

### Steps to reproduce

1. Reduce the size of the runner used in the `oci` step
2. Run the step (in a fork, for example)
3. Observe the runner running out of memory

### Expected behavior

The `oci` GHA job should not fail

### Environment

- OS: [e.g. Linux, MacOS]
- Shell [e.g. bash, zsh, powershell]
- wasmcloud Version [e.g. 0.81.0]

### Screenshots / Logs / Additional context

![image](https://github.com/user-attachments/assets/d3837572-55a5-4dd3-841b-255979fac61f)

**Comment by brooksmtownsend** [comment]
@vados-cosmonic could this be simply done today by adding `--no-default-features` to the oci build step? Since the providers are essentially feature gated in the Cargo.toml (if this is true, this issue becomes a good first issue 🤔 )

**Comment by vados-cosmonic** [comment]
Yeah unfortunately it's not that easy! there is no "oci build step" -- in the quest for efficiency we built one step that builds everything (and then other steps depend on that one step) -- so we need to actually modify the existing workflow/untangle it.

Now, the reason that everything gets built together in the first place is the Cargo config, and that's the reason we have to do the top level `wash build -p src/bin/...` in the first place rather than building right from a given provider's directory. Definitely some thought required here on what we should change!

**Comment by thomastaylor312** [comment]
Just noting that #3739 started the process of making this possible

**Comment by brooksmtownsend** [comment]
I've spoken to @rvolosatovs a bit about this, and we should be able to do this with nix outputs after #3739. Along with the latest changes to wash and wasmCloud successful checks, we should be able to change our CI to only rebuild capability providers if their contents (or the top level Cargo.toml/lock) change.

**Comment by brooksmtownsend** [comment]
Done after https://github.com/wasmCloud/wasmCloud/pull/4018!
